src/vstorage-query.ts

```diff
@@ -1,7 +1,67 @@
-import { abciQuery } from './abci-query';
+import { timeoutDurationMS } from './constants';
 
-export const vstorageQuery = async (
-  node: string,
-  path: string,
-  blockHeight?: number
-) => abciQuery(node, `/custom/vstorage/data/${path}`, blockHeight);
+const getNodeIterator = (nodes: string[]) => {
+  let currentIndex = Math.floor(Math.random() * nodes.length);
+  const next = () => {
+    currentIndex += 1;
+    if (currentIndex >= nodes.length) {
+      currentIndex = 0;
+    }
+    return nodes[currentIndex];
+  };
+
+  return { next };
+};
+
+/**
+ * Queries a vstorage path from a random node from `nodes`. Every
+ * `timeoutDurationMS`, it will queue up another request to a new node until
+ * all nodes are exhausted, returning the first result it gets.
+ *
+ * @param nodes The API nodes to request vstorage data from
+ * @param path The vstorage path
+ * @returns The vstorage query result
+ */
+export const vstorageQuery = async (nodes: string[], path: string) => {
+  const nodeIterator = getNodeIterator(nodes);
```

Comment:
Consider moving "twice through" logic here.
```suggestion
  const nodesRepeatedOnce = [...nodes, ...nodes];
  const nodeIterator = getNodeIterator(nodesRepeatedOnce);
```

Reply:
Did something different here, but similar effect